File: pset6/sentimental-readability/readability.py
from cs50 import get_string


def main():
    # Prompt user for text
    text = get_string("Text: ")

    # Count number of letters, words and strings in text
    letter_count = count_letters(text)
    word_count = count_words(text)
    sentence_count = count_sentences(text)
    if letter_count > 1:
        word_count += 1

    # Calculate Coleman-Liau Index, return reading level to terminal
    coleman_liau(letter_count, word_count, sentence_count)


def count_letters(text):
    """ Calculate number of letters in a string """
    n = len(text)
    letter_count = 0
    for i in range(n):
        if ((ord(text[i]) >= ord('A') and ord(text[i]) <= ord('Z'))
                or (ord(text[i]) >= ord('a') and ord(text[i]) <= ord('z'))):
            letter_count += 1
    return letter_count


def count_words(text):
    """ Calculate number of words in a string """
    n = len(text)
    word_count = 0
    for i in range(n):
        if text[i] == ' ':
            word_count += 1
    return word_count


def count_sentences(text):
    """ Calculate number of sentences in a string """
    sentence_count = 0
    n = len(text)
    for i in range(n):
        if text[i] == '!' or text[i] == '.' or text[i] == '?':
            sentence_count += 1
    return sentence_count


def coleman_liau(letter_count, word_count, sentence_count):
    """ Calculate CL index, round to nearest integer
    and map integer to corresponding reading level"""
    index = (0.0588 * 100 * letter_count / word_count) - (0.296 * 100 * sentence_count / word_count) - 15.8
    if index < 1:
        print("Before Grade 1")
    elif index >= 16:
        print("Grade 16+")
    else:
        int(round(index))
        print("Grade " + str(round(index)))


if __name__ == "__main__":
    main()

File: pset2/readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
float coleman_liau(string text);


int main(void)
{
    // Prompt user for text
    string text = get_string("Text: ");

    // Calculate CL Index, round to obtain reading level
    int reading_level = (int) round(coleman_liau(text));
    if (reading_level >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (reading_level < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", reading_level);
    }
}

// Count number of letters in given text
int count_letters(string text)
{
    int sum = 0;

    // Add 1 letter for every ascii-mapped character present
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (toupper(text[i]) >= 65 && toupper(text[i]) <= 90)
        {
            sum += 1;
        }
    }
    return sum;
}

// Count number of words in given text
int count_words(string text)
{
    int sum = 0;

    // If even one letter exists, then one word exists
    if (text[0] != '\0')
    {
        sum += 1;
    }

    // Add 1 word for every space present
    for (int i = 0, n = strlen(text); i < n; i ++)
    {
        if (text[i] == ' ')
        {
            sum += 1;
        }
    }

    return sum;
}

// Count number of sentences in given text
int count_sentences(string text)
{
    int sum = 0;
    for (int i = 0, n = strlen(text); i < n; i ++)
    {
        // Add 1 sentence for every punctuation character present
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sum += 1;
        }
    }
    return sum;
}

// Calculate Coleman-Liau Index
float coleman_liau(string text)
{
    float letterCount = (float) count_letters(text);
    float wordCount = (float) count_words(text);
    float sentenceCount = (float) count_sentences(text);
    float cl_ind = (0.0588 * ((letterCount / wordCount) * 100)) - (0.296 * ((sentenceCount / wordCount) * 100)) - 15.8;
    return cl_ind;
}